README.md
![cf](http://i.imgur.com/7v5ASc8.png) http-simple-persistence
====

## API server for books, by Dave and David

##### To install, clone this repo then run

    npm install
    ./index.js [port]

##### Available endpoints and methods:
- GET
  - /books: gets a list of available books
  - /books/resource[/resource]: returns all data for each item listed.
- POST
  - /books: adds a new book to the database
- PUT
  - /books/resource: replaces properties of the specified resource with information provided
- DELETE
  - /books/resource: deleted the specified resource

##### Data is sent and received in JSON format, as follows:

    {
      "title" : "Gone With The Wind",
      "pub_year" : 1964,
      "author" : "Margaret Mitchell",
      "genre" : "fiction",
      "read" : false,
      "resource" : "gone_with_the_wind_1964"
    }

  title and pub_year are required as they are combined to create the resource ID.

##### Tests are available:

    npm test


CHANGELOG.md
## changelog for http-simple-persistence


### 1.0.0 (2016-05-18)

#### changes:
  - Full CRUD operations in place via JSON files
  - supports GET, POST, PUT, and DELETE requests.


test/db.test.js
const db = require ('../db.js');
const assert = require('chai').assert;


describe('books module',() => {

  describe('on create', () =>{

    it('writes to json file and returns an object', ( done ) =>{
      db.create({title:'herbook', pub_year:'1986'})
      .then(data => {
        assert.isOk(data);
        done();
      })
      .catch(error =>{
        assert.isOk(error);
        done();
      });
    });

  });

  describe('on read', () =>{

    it('returns an array of objects when given an array of resources', ( done ) =>{
      db.read(['herbook_1986'])
      .then( data =>{
        assert.equal(data.length, 2);
        done();
      })
      .catch( error =>{
        assert.isOk(error);
        done();
      });

    });


    it('returns an array of resources when given an empty array', (done) =>{
      db.read([])
      .then( data =>{
        assert.isOk(data);
        done();
      })
      .catch( error =>{
        assert.isOk(error);
        done();
      });
    });
  });

  describe('on update', () =>{

    it('renames a filename and resource and returns an object', (done) =>{
      db.update('herbook_1986', {title:'herbook', pub_year:'2008', resource:'herbook_2008'})
      .then( data => {
        assert.isOk(data);
        done();
      })
      .catch( err => {
        assert.isOk(err);
        done();
      });
    });

  });

  describe('on delete', () =>{

    it('returns a delete message', (done) =>{
      db.delete('herbook_2008')
      .then( data => {
        assert.equal(data.message, 'deleted herbook_2008');
        done();
      })
      .catch( err => {
        assert.isOk(err);
        done();
      });
    });

  });
});


test/server.test.js
const server = require('../server');
const chai = require('chai');
const chaiHttp = require('chai-http');
const assert = chai.assert;
chai.use(chaiHttp);

describe('server',() => {
  const request = chai.request(server);
  const testGetData =  {
    'title' : 'Time Enough For Love',
    'author' : 'Robert Heinlein',
    'genre' : 'Science Fiction',
    'pub_year' : 1973,
    'read' : true
  };
  const testPostData =  {
    'title' : 'Gone With The Wind',
    'author' : 'Margaret Mitchell',
    'genre' : 'fiction',
    'pub_year' : 1964,
    'read' : false
  };
  const testPutData = {
    'title' : 'Gone With The Wind',
    'author' : 'Margaret Mitchell',
    'genre' : 'fiction',
    'pub_year' : 1964,
    'read' : false,
    'resource' : 'Gone_With_The_Wind_1964'
  };
  const testResource = 'Gone_With_The_Wind_1964';


  describe('GET',() => {
    describe('error handling',() => {
      it('without a valid endpoint returns a 404 status code', done => {
        request
          .get('/test')
          .end((err,res) => {
            assert.equal(res.statusCode, 404);
            assert.ok(res.text);
            done();
          });
      });
    });

    describe('operations', () => {

      // populating DB with data to retrieve
      before(done => {
        request
          .post('/books')
          .set('Content-Type', 'application/json')
          .send(JSON.stringify(testGetData))
          .end(() => done());
      });

      it('/books returns a list of available resources', done => {
        request
          .get('/books')
          .end((err,res) => {
            assert.equal(res.statusCode, 200);
            assert.propertyVal(res.header,'content-type','application/json');
            let results = JSON.parse(res.text);
            assert.isArray(results);
            assert.isAtLeast(results.length, 1);
            done();
          });
      });

      it('/books/resource returns the data for that resource', done => {
        request
          .get('/books/gone_with_the_wind_1964')
          .end((err,res) => {
            assert.equal(res.statusCode, 200);
            assert.propertyVal(res.header,'content-type','application/json');
            assert.isArray(JSON.parse(res.text));
            done();
          });
      });
    });
  });

  describe('POST',() => {
    describe('error handling',() => {
      it('without a valid endpoint returns a 400 status code', done => {
        request
          .post('/test')
          .end((err,res) => {
            assert.equal(res.statusCode, 400);
            assert.propertyVal(res.header,'content-type','text/plain');
            assert.ok(res.text);
            done();
          });
      });
    });
    describe('operations', () => {
      it('/books returns JSON data with resource added', done => {
        request
          .post('/books')
          .set('Content-Type', 'application/json')
          .send(JSON.stringify(testPostData))
          .end((err,res) => {
            var result = JSON.parse(res.text);
            assert.equal(res.statusCode, 201);
            assert.isObject(result);
            assert.property(result, 'resource');
            assert.propertyVal(res.header,'content-type','application/json');
            assert.propertyVal(result, 'genre', testPostData.genre);
            done();
          });
      });
    });
  });

  describe('PUT',() => {
    describe('error handling',() => {
      it('without a valid endpoint returns a 400 status code', done => {
        request
          .put('/test')
          .end((err,res) => {
            assert.equal(res.statusCode, 400);
            assert.propertyVal(res.header,'content-type','text/plain');
            assert.ok(res.text);
            done();
          });
      });
    });

    describe('operations', () => {
      it('/books/resource returns the updated item', done => {
        request
          .put('/books/'+testResource)
          .set('Content-Type', 'application/json')
          .send(JSON.stringify(testPutData))
          .end((err,res) => {
            var result = JSON.parse(res.text);
            assert.equal(res.statusCode, 201);
            assert.propertyVal(res.header,'content-type','application/json');
            assert.isObject(result);
            assert.property(result, 'resource');
            done();
          });
      });
    });
  });

  describe('DELETE',() => {
    describe('error handling',() => {
      it('without a valid endpoint returns a 400 status code.', done => {
        request
          .delete('/test')
          .end((err,res) => {
            assert.equal(res.statusCode, 400);
            assert.ok(res.text);
            done();
          });
      });
    });

    describe('operations', () => {
      it('/books/resource returns the deletion notification', done => {
        request
          .delete('/books/'+testResource)
          .set('Content-Type', 'application/json')
          .end((err,res) => {
            var result = JSON.parse(res.text);
            assert.equal(res.statusCode, 201);
            assert.isOk(result);
            done();
          });
      });
    });
  });

  describe('invalid method',() => {

    it('sending PATCH request generates 405 error', done => {
      request
        .patch('/books')
        .end((err,res) => {
          assert.equal(res.statusCode, 405);
          assert.propertyVal(res.header,'content-type','text/plain');
          assert.ok(res.text);
          done();
        });
    });
  });




  after(done => {
    server.close(done);
  });

});


db.js
const sander = require('sander');

const books = Object.create(null);
const dir = 'data';
const path = dir + '/';

books.create = function(obj){

  obj.resource = generateFileName(obj);
  const newPath = path + obj.resource + '.json';
  const objJson = JSON.stringify(obj);

  return sander.writeFile(newPath, objJson)
  .then( () =>{
    return obj;
  });
};

books.read = function(resourceArray) {

  if (resourceArray.length < 1) {
    // Given an empty array, returns a list of resources
    return sander.readdir(path).
    then( fileNames => fileNames.map( e => e.slice(0, -5)));
  } else {
    return sander.readdir(path)
    .then( totalArray => {
      return totalArray.filter(item => {
        var inArray = false;
        for (var x = 0; x < resourceArray.length; x++){
          if (resourceArray[x] === item.slice(0, -5)) {
            inArray = true;
          }
        }
        return inArray;
      });
    })
    .then(resourceArray => resourceArray.map( e => path + e))
    .then(filePaths => {
      return filePaths.map( p => {
        return sander.readFile(p, {encoding: 'utf-8'});
      });
    })
    .then(filePromises => Promise.all(filePromises))
    .then(jsonItems => jsonItems.map( j => JSON.parse(j)));
  }
};

books.update = function(resource, obj){

  const origPath = path + resource + '.json';
  const newPath = path + obj.resource + '.json';
  const objJson = JSON.stringify(obj);

  return sander.writeFile(newPath, objJson)
  .then(sander.unlink(origPath))
  .then( () =>{
    return obj;
  });

};

books.delete = function(resource){

  const origPath = path + resource + '.json';
  return sander.unlink(origPath)
  .then( () => {
    return {message:'deleted ' + resource};
  });
};

function generateFileName(obj){
  return obj.title.split(' ').join('_') + '_' + obj.pub_year;
}

module.exports = books;


server.js
const http = require('http');
const url = require('url');
const db = require('./db');

const server = http.createServer((req,res) => {
  const requestPath = url.parse(req.url, true).pathname;
  const validEndPoint = /^\/books/.test(requestPath);

  switch (req.method) {

  case 'GET': {
    if (validEndPoint) {
      const resources = requestPath.split(/[\\/]/).splice(2);
      db.read(resources)
        .then(data => {
          writeJson(data, 200);
        })
        .catch(err => {
          writeError(err, 404);
        });
    } else {
      res.writeHead(404, {'Content-Type': 'text/plain'});
      res.write('Not Found.\n\nSorry, what you were looking for is not available.');
      res.end();
    }
    break;
  }

  case 'POST': {
    if (validEndPoint) {
      let body='';
      req.on('data', (chunk) => body += chunk);
      req.on('end', () => {
        db.create(JSON.parse(body))
          .then(data => {
            writeJson(data, 201);
          })
          .catch(err => {
            writeError(err, 400);
          });
      });
    } else {
      badRequest();
    }
    break;
  }

  case 'PUT': {
    if (validEndPoint) {
      const resource = requestPath.split(/[\\/]/).splice(1)[1];
      let body='';
      req.on('data', (chunk) => body += chunk);
      req.on('end', () => {
        db.update(resource, JSON.parse(body))
          .then( data => {
            writeJson(data, 201);
          })
          .catch(err => {
            writeError(err, 400);
          });
      });

    } else {
      badRequest();
    }
    break;
  }

  case 'DELETE': {
    if (validEndPoint) {
      const resource = requestPath.split(/[\\/]/).splice(1)[1];
      db.delete(resource)
        .then(data => {
          writeJson(data, 201);
        })
        .catch(err => {
          writeError(err, 400);
        });

    } else {
      badRequest();
    }
    break;
  }

  default: {
    res.writeHead(405, {'Content-Type': 'text/plain'});
    res.write('Method not supported.');
    res.end();
  }
  }

  function badRequest() {
    res.writeHead(400, {'Content-Type': 'text/plain'});
    res.write('Bad Request.\n\nSorry, that request is not supported');
    res.end();
  }
  function writeJson(data, code) {
    res.writeHead(code, {'Content-Type': 'application/json'});
    res.write(JSON.stringify(data));
    res.end();
  }
  function writeError(err, code) {
    res.writeHead(code, {'Content-Type': 'text/plain'});
    res.write(err);
    res.end();
  }
});

module.exports = server;
